software/runtime/kernel/mempool_cholesky_q32p.h

```diff
@@ -0,0 +1,996 @@
+// Copyright 2021 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Author: Marco Bertuletti, ETH Zurich
+
+#include "kernel/mempool_sqrt_q32s.h"
+
+void mempool_cholesky_q32p(int32_t *pSrc, int32_t *pL, const uint32_t n) {
+
+  int32_t sum;
+  int32_t pivot, diag;
+  uint32_t i, j, k;
+  int32_t a0, a1, a2, a3;
+  int32_t b0, b1, b2, b3;
+
+  uint32_t absolute_core_id = mempool_get_core_id();
+  uint32_t num_cores = mempool_get_core_count();
+  uint32_t row_id;
+  if (absolute_core_id % n == 0)
+    row_id = absolute_core_id / n;
+  else
+    row_id = n + 1;
+
+  for (j = 0; j < n; j++) {
+
+    /* Elements on the diagonal are computed with a single core */
+    if (row_id == (j % num_cores)) {
+      pivot = pSrc[j * n + j];
+      sum = 0;
+      for (k = 0; k < 4 * (j >> 2U); k++) {
+        a0 = pL[j * n + k];
+        a1 = pL[j * n + k + 1];
+        a2 = pL[j * n + k + 2];
+        a3 = pL[j * n + k + 3];
+        asm volatile("mul  %[a0],%[a0],%[a0];"
+                     "mul  %[a1],%[a1],%[a1];"
+                     "mul  %[a2],%[a2],%[a2];"
+                     "mul  %[a3],%[a3],%[a3];"
+                     "addi %[a0],%[a0],%[h];"
+                     "addi %[a1],%[a1],%[h];"
+                     "addi %[a2],%[a2],%[h];"
+                     "addi %[a3],%[a3],%[h];"
+                     "srai  %[a0],%[a0],%[s];"
+                     "srai  %[a1],%[a1],%[s];"
+                     "srai  %[a2],%[a2],%[s];"
+                     "srai  %[a3],%[a3],%[s];"
+                     "add  %[sum],%[a0],%[sum];"
+                     "add  %[sum],%[a1],%[sum];"
+                     "add  %[sum],%[a2],%[sum];"
+                     "add  %[sum],%[a3],%[sum];"
+                     : [a0] "+&r"(a0), [a1] "+&r"(a1), [a2] "+&r"(a2),
+                       [a3] "+&r"(a3), [sum] "+&r"(sum)
+                     : [s] "I"(FIXED_POINT), [h] "I"(HALF)
+                     :);
+      }
+      switch (j % 4) {
+      case 3:
+        a0 = pL[j * n + k];
+        a1 = pL[j * n + k + 1];
+        a2 = pL[j * n + k + 2];
+        asm volatile(
+            "mul  %[a0],%[a0],%[a0];"
+            "mul  %[a1],%[a1],%[a1];"
+            "mul  %[a1],%[a2],%[a2];"
```

Comment:
```suggestion
            "mul  %[a2],%[a2],%[a2];"
```
Another small nitpick, this should probably be `a2`. Then it is exactly the same as in `..._q32s.h`. So probably also for the single vs multicore implementations, maybe you can use the parallelized ones also for the single core by just setting num_cores to 1 and then skipping any barriers?